File: scripts/SELemmas.jl
#!/usr/bin/env -S julia --startup-file=no

using BFloat16s: BFloat16

push!(LOAD_PATH, @__DIR__)
using FloatSummaries: summarize_se, unsafe_exponent,
    normal_summaries, normal_two_sum_summaries,
    EFTSummary, LemmaVerifier, LemmaHelper


const ± = false:true


function verify_two_sum_se_lemmas(
    ::Type{T},
    summaries::Vector{EFTSummary},
) where {T}

    p = precision(T)
    pos_zero = summarize_se(+zero(T))
    neg_zero = summarize_se(-zero(T))
    possible_inputs = normal_summaries(summarize_se, T)
    add_case! = LemmaHelper(T)
    lemma_dict = Dict{String,Int}()

    for x in possible_inputs, y in possible_inputs

        verifier = LemmaVerifier(summaries, x, y, lemma_dict)
        sx = signbit(x)
        ex = unsafe_exponent(x)
        sy = signbit(y)
        ey = unsafe_exponent(y)
        x_zero = (x == pos_zero) | (x == neg_zero)
        y_zero = (y == pos_zero) | (y == neg_zero)
        same_sign = (sx == sy)
        diff_sign = (sx != sy)

        if x_zero | y_zero ############################### LEMMA FAMILY SE-Z (2)

            # Lemmas in Family SE-Z (for "zero") apply
            # when one or both addends are zero.

            # Lemma SE-Z1: Both addends are zero.
            verifier("SE-Z1-PP", (x == pos_zero) & (y == pos_zero)) do lemma
                add_case!(lemma, pos_zero, pos_zero)
            end
            verifier("SE-Z1-PN", (x == pos_zero) & (y == neg_zero)) do lemma
                add_case!(lemma, pos_zero, pos_zero)
            end
            verifier("SE-Z1-NP", (x == neg_zero) & (y == pos_zero)) do lemma
                add_case!(lemma, pos_zero, pos_zero)
            end
            verifier("SE-Z1-NN", (x == neg_zero) & (y == neg_zero)) do lemma
                add_case!(lemma, neg_zero, pos_zero)
            end

            # Lemma SE-Z2: One addend is zero.
            verifier("SE-Z2-X", y_zero & !x_zero) do lemma
                add_case!(lemma, x, pos_zero)
            end
            verifier("SE-Z2-Y", x_zero & !y_zero) do lemma
                add_case!(lemma, y, pos_zero)
            end

        else #################################################### NONZERO LEMMAS

            # From this point forward, all lemma statements carry
            # an implicit assumption that both addends are nonzero.

            ##################################################### LEMMA SE-I (1)

            # Lemma SE-I (for "identical") applies to addends
            # returned unchanged by the TwoSum algorithm.

            verifier("SE-I-X", (ex > ey + (p+1)) | ((ex == ey + (p+1)) & same_sign)) do lemma
                add_case!(lemma, x, y)
            end
            verifier("SE-I-Y", (ey > ex + (p+1)) | ((ey == ex + (p+1)) & same_sign)) do lemma
                add_case!(lemma, y, x)
            end

            ############################################## LEMMA FAMILY SE-S (5)

            # Lemmas in Family SE-S apply to addends with the same sign.

            verifier("SE-S1-X", same_sign & (ex == ey + p)) do lemma
                add_case!(lemma, (sx, ex:ex+1), (!sy, ey-(p-1):ex-p))
                add_case!(lemma, x            , y                   )
            end
            verifier("SE-S1-Y", same_sign & (ey == ex + p)) do lemma
                add_case!(lemma, (sy, ey:ey+1), (!sx, ex-(p-1):ey-p))
                add_case!(lemma, y            , x                   )
            end

            verifier("SE-S2-X", same_sign & (ex == ey + (p-1))) do lemma
                add_case!(lemma, (sx, ex:ex+1), pos_zero          )
                add_case!(lemma, (sx, ex:ex+1), (±, ey-(p-1):ex-p))
            end
            verifier("SE-S2-Y", same_sign & (ey == ex + (p-1))) do lemma
                add_case!(lemma, (sy, ey:ey+1), pos_zero          )
                add_case!(lemma, (sy, ey:ey+1), (±, ex-(p-1):ey-p))
            end

            verifier("SE-S3-X", same_sign & (ex == ey + (p-2))) do lemma
                add_case!(lemma, (sx, ex:ex+1), pos_zero                )
                add_case!(lemma, (sx, ex:ex+1), (!sy, ey-(p-1):ex-p    ))
                add_case!(lemma, (sx, ex     ), ( sy, ey-(p-1):ex-p    ))
                add_case!(lemma, (sx, ex+1   ), ( sy, ey-(p-1):ex-(p-1)))
            end
            verifier("SE-S3-Y", same_sign & (ey == ex + (p-2))) do lemma
                add_case!(lemma, (sy, ey:ey+1), pos_zero                )
                add_case!(lemma, (sy, ey:ey+1), (!sx, ex-(p-1):ey-p    ))
                add_case!(lemma, (sy, ey     ), ( sx, ex-(p-1):ey-p    ))
                add_case!(lemma, (sy, ey+1   ), ( sx, ex-(p-1):ey-(p-1)))
            end

            verifier("SE-S4-X", same_sign & (ex > ey) & (ex < ey + (p-2))) do lemma
                add_case!(lemma, (sx, ex:ex+1), pos_zero              )
                add_case!(lemma, (sx, ex     ), (±, ey-(p-1):ex-p    ))
                add_case!(lemma, (sx, ex+1   ), (±, ey-(p-1):ex-(p-1)))
            end
            verifier("SE-S4-Y", same_sign & (ey > ex) & (ey < ex + (p-2))) do lemma
                add_case!(lemma, (sy, ey:ey+1), pos_zero              )
                add_case!(lemma, (sy, ey     ), (±, ex-(p-1):ey-p    ))
                add_case!(lemma, (sy, ey+1   ), (±, ex-(p-1):ey-(p-1)))
            end

            verifier("SE-S5", same_sign & (ex == ey)) do lemma
                add_case!(lemma, (sx, ex+1), pos_zero)
                add_case!(lemma, (sx, ex+1), (±, ex-(p-1)))
            end

            ############################################## LEMMA FAMILY SE-D (5)

            # Lemmas in Family SE-D apply to addends with different signs.

            verifier("SE-D1-X", diff_sign & (ex == ey + (p+1))) do lemma
                add_case!(lemma, (sx, ex-1), (!sy, ey-(p-1):ex-(p+2)))
                add_case!(lemma, x         , y                       )
            end
            verifier("SE-D1-Y", diff_sign & (ey == ex + (p+1))) do lemma
                add_case!(lemma, (sy, ey-1), (!sx, ex-(p-1):ey-(p+2)))
                add_case!(lemma, y         , x                       )
            end

            verifier("SE-D2-X", diff_sign & (ex == ey + p)) do lemma
                add_case!(lemma, (sx, ex-1), pos_zero                )
                add_case!(lemma, (sx, ex-1), ( sy, ey-(p-1):ex-(p+2)))
                add_case!(lemma, (sx, ex-1), (!sy, ey-(p-1):ex-(p+1)))
                add_case!(lemma, (sx, ex  ), (!sy, ey-(p-1):ex-p    ))
                add_case!(lemma, x         , y                       )
            end
            verifier("SE-D2-Y", diff_sign & (ey == ex + p)) do lemma
                add_case!(lemma, (sy, ey-1), pos_zero                )
                add_case!(lemma, (sy, ey-1), ( sx, ex-(p-1):ey-(p+2)))
                add_case!(lemma, (sy, ey-1), (!sx, ex-(p-1):ey-(p+1)))
                add_case!(lemma, (sy, ey  ), (!sx, ex-(p-1):ey-p    ))
                add_case!(lemma, y         , x                       )
            end

            verifier("SE-D3-X", diff_sign & (ex > ey + 1) & (ex < ey + p)) do lemma
                add_case!(lemma, (sx, ex-1:ex), pos_zero              )
                add_case!(lemma, (sx, ex-1   ), (±, ey-(p-1):ex-(p+1)))
                add_case!(lemma, (sx, ex     ), (±, ey-(p-1):ex-p    ))
            end
            verifier("SE-D3-Y", diff_sign & (ey > ex + 1) & (ey < ex + p)) do lemma
                add_case!(lemma, (sy, ey-1:ey), pos_zero              )
                add_case!(lemma, (sy, ey-1   ), (±, ex-(p-1):ey-(p+1)))
                add_case!(lemma, (sy, ey     ), (±, ex-(p-1):ey-p    ))
            end

            verifier("SE-D4-X", diff_sign & (ex == ey + 1)) do lemma
                add_case!(lemma, (sx, ex-p:ex), pos_zero )
                add_case!(lemma, (sx, ex     ), (±, ex-p))
            end
            verifier("SE-D4-Y", diff_sign & (ey == ex + 1)) do lemma
                add_case!(lemma, (sy, ey-p:ey), pos_zero )
                add_case!(lemma, (sy, ey     ), (±, ey-p))
            end

            verifier("SE-D5", diff_sign & (ex == ey)) do lemma
                add_case!(lemma, pos_zero          , pos_zero)
                add_case!(lemma, (±, ex-(p-1):ex-1), pos_zero)
            end

        end
        @assert isone(verifier.count[])
    end

    println(T, " TwoSum-SE lemmas:")
    for (name, count) in sort!(collect(lemma_dict))
        println("    ", name, ": ", count)
    end

    return nothing
end


if !isfile("Float16TwoSumSE.bin")
    open("Float16TwoSumSE.bin", "w") do io
        write(io, normal_two_sum_summaries(summarize_se, Float16))
    end
end
@assert isfile("Float16TwoSumSE.bin")
@assert filesize("Float16TwoSumSE.bin") == 38_638 * sizeof(EFTSummary)
const FLOAT16_TWO_SUM_SE = Vector{EFTSummary}(undef, 38_638)
read!("Float16TwoSumSE.bin", FLOAT16_TWO_SUM_SE)
verify_two_sum_se_lemmas(Float16, FLOAT16_TWO_SUM_SE)
println("Verified all Float16 TwoSum-SE lemmas.")
flush(stdout)


if !isfile("BFloat16TwoSumSE.bin")
    open("BFloat16TwoSumSE.bin", "w") do io
        write(io, normal_two_sum_summaries(summarize_se, BFloat16))
    end
end
@assert isfile("BFloat16TwoSumSE.bin")
@assert filesize("BFloat16TwoSumSE.bin") == 548_026 * sizeof(EFTSummary)
const BFLOAT16_TWO_SUM_SE = Vector{EFTSummary}(undef, 548_026)
read!("BFloat16TwoSumSE.bin", BFLOAT16_TWO_SUM_SE)
verify_two_sum_se_lemmas(BFloat16, BFLOAT16_TWO_SUM_SE)
println("Verified all BFloat16 TwoSum-SE lemmas.")
flush(stdout)

File: scripts/se_lemmas.py
import z3
from smt_utils import BoolVar, IntVar, FloatVar
from typing import Callable


def two_sum_se_lemmas(
    x: FloatVar,
    y: FloatVar,
    s: FloatVar,
    e: FloatVar,
    sx: BoolVar,
    sy: BoolVar,
    ss: BoolVar,
    se: BoolVar,
    ex: IntVar,
    ey: IntVar,
    es: IntVar,
    ee: IntVar,
    is_zero: Callable[[FloatVar], z3.BoolRef],
    is_positive: Callable[[FloatVar], z3.BoolRef],
    is_negative: Callable[[FloatVar], z3.BoolRef],
    is_equal: Callable[[FloatVar, FloatVar], z3.BoolRef],
    p: IntVar,
    one: IntVar,
    two: IntVar,
) -> dict[str, z3.BoolRef]:

    result: dict[str, z3.BoolRef] = {}

    x_zero: z3.BoolRef = is_zero(x)
    x_pos_zero: z3.BoolRef = z3.And(is_positive(x), x_zero)
    x_neg_zero: z3.BoolRef = z3.And(is_negative(x), x_zero)

    y_zero: z3.BoolRef = is_zero(y)
    y_pos_zero: z3.BoolRef = z3.And(is_positive(y), y_zero)
    y_neg_zero: z3.BoolRef = z3.And(is_negative(y), y_zero)
    xy_nonzero: z3.BoolRef = z3.And(z3.Not(x_zero), z3.Not(y_zero))

    s_zero: z3.BoolRef = is_zero(s)
    s_pos_zero: z3.BoolRef = z3.And(is_positive(s), s_zero)
    s_neg_zero: z3.BoolRef = z3.And(is_negative(s), s_zero)
    e_pos_zero: z3.BoolRef = z3.And(is_positive(e), is_zero(e))

    s_equals_x: z3.BoolRef = is_equal(s, x)
    s_equals_y: z3.BoolRef = is_equal(s, y)
    e_equals_x: z3.BoolRef = is_equal(e, x)
    e_equals_y: z3.BoolRef = is_equal(e, y)

    same_sign: z3.BoolRef = sx == sy
    diff_sign: z3.BoolRef = sx != sy

    ###################################################### LEMMA FAMILY SE-Z (2)

    # Lemmas in Family SE-Z (for "zero") apply
    # when one or both addends are zero.

    # Lemma SE-Z1: Both addends are zero.
    result["TwoSum-SE-Z1-PP"] = z3.Implies(
        z3.And(x_pos_zero, y_pos_zero),
        z3.And(s_pos_zero, e_pos_zero),
    )
    result["TwoSum-SE-Z1-PN"] = z3.Implies(
        z3.And(x_pos_zero, y_neg_zero),
        z3.And(s_pos_zero, e_pos_zero),
    )
    result["TwoSum-SE-Z1-NP"] = z3.Implies(
        z3.And(x_neg_zero, y_pos_zero),
        z3.And(s_pos_zero, e_pos_zero),
    )
    result["TwoSum-SE-Z1-NN"] = z3.Implies(
        z3.And(x_neg_zero, y_neg_zero),
        z3.And(s_neg_zero, e_pos_zero),
    )

    # Lemma SE-Z2: One addend is zero.
    result["TwoSum-SE-Z2-X"] = z3.Implies(
        z3.And(y_zero, z3.Not(x_zero)),
        z3.And(s_equals_x, e_pos_zero),
    )
    result["TwoSum-SE-Z2-Y"] = z3.Implies(
        z3.And(x_zero, z3.Not(y_zero)),
        z3.And(s_equals_y, e_pos_zero),
    )

    ############################################################# LEMMA SE-I (1)

    # Lemma SE-I (for "identical") applies to addends
    # returned unchanged by the TwoSum algorithm.

    result["TwoSum-SE-I-X"] = z3.Implies(
        z3.And(
            xy_nonzero,
            z3.Or(ex > ey + (p + one), z3.And(ex == ey + (p + one), same_sign)),
        ),
        z3.And(s_equals_x, e_equals_y),
    )
    result["TwoSum-SE-I-Y"] = z3.Implies(
        z3.And(
            xy_nonzero,
            z3.Or(ey > ex + (p + one), z3.And(ey == ex + (p + one), same_sign)),
        ),
        z3.And(s_equals_y, e_equals_x),
    )

    ########################################################### HELPER FUNCTIONS

    def se_case(
        ss_value: BoolVar | tuple[BoolVar] | None,
        es_range: IntVar | tuple[IntVar, IntVar],
        se_value: BoolVar | tuple[BoolVar] | None,
        ee_range: IntVar | tuple[IntVar, IntVar],
    ) -> z3.BoolRef:
        result: list[z3.BoolRef] = []
        if isinstance(ss_value, tuple):
            result.append(ss != ss_value[0])
        elif ss_value is not None:
            result.append(ss == ss_value)
        if isinstance(es_range, tuple):
            result.append(es >= es_range[0])
            result.append(es <= es_range[1])
        else:
            result.append(es == es_range)
        if isinstance(se_value, tuple):
            result.append(se != se_value[0])
        elif se_value is not None:
            result.append(se == se_value)
        if isinstance(ee_range, tuple):
            result.append(ee >= ee_range[0])
            result.append(ee <= ee_range[1])
        else:
            result.append(ee == ee_range)
        return z3.And(*result)

    def se_case_zero(
        ss_value: BoolVar | tuple[BoolVar] | None,
        es_range: IntVar | tuple[IntVar, IntVar],
    ) -> z3.BoolRef:
        result: list[z3.BoolRef] = []
        if isinstance(ss_value, tuple):
            result.append(ss != ss_value[0])
        elif ss_value is not None:
            result.append(ss == ss_value)
        if isinstance(es_range, tuple):
            result.append(es >= es_range[0])
            result.append(es <= es_range[1])
        else:
            result.append(es == es_range)
        result.append(e_pos_zero)
        return z3.And(*result)

    ###################################################### LEMMA FAMILY SE-S (5)

    # Lemmas in Family SE-S apply to addends with the same sign.

    result["TwoSum-SE-S1-X"] = z3.Implies(
        z3.And(same_sign, ex == ey + p),
        z3.Or(
            se_case(sx, (ex, ex + one), (sy,), (ey - (p - one), ex - p)),
            z3.And(s_equals_x, e_equals_y),
        ),
    )
    result["TwoSum-SE-S1-Y"] = z3.Implies(
        z3.And(same_sign, ey == ex + p),
        z3.Or(
            se_case(sy, (ey, ey + one), (sx,), (ex - (p - one), ey - p)),
            z3.And(s_equals_y, e_equals_x),
        ),
    )

    result["TwoSum-SE-S2-X"] = z3.Implies(
        z3.And(same_sign, ex == ey + (p - one)),
        z3.Or(
            se_case_zero(sx, (ex, ex + one)),
            se_case(sx, (ex, ex + one), None, (ey - (p - one), ex - p)),
        ),
    )
    result["TwoSum-SE-S2-Y"] = z3.Implies(
        z3.And(same_sign, ey == ex + (p - one)),
        z3.Or(
            se_case_zero(sy, (ey, ey + one)),
            se_case(sy, (ey, ey + one), None, (ex - (p - one), ey - p)),
        ),
    )

    result["TwoSum-SE-S3-X"] = z3.Implies(
        z3.And(same_sign, ex == ey + (p - two)),
        z3.Or(
            se_case_zero(sx, (ex, ex + one)),
            se_case(sx, (ex, ex + one), (sy,), (ey - (p - one), ex - p)),
            se_case(sx, ex, sy, (ey - (p - one), ex - p)),
            se_case(sx, ex + one, sy, (ey - (p - one), ex - (p - one))),
        ),
    )
    result["TwoSum-SE-S3-Y"] = z3.Implies(
        z3.And(same_sign, ey == ex + (p - two)),
        z3.Or(
            se_case_zero(sy, (ey, ey + one)),
            se_case(sy, (ey, ey + one), (sx,), (ex - (p - one), ey - p)),
            se_case(sy, ey, sx, (ex - (p - one), ey - p)),
            se_case(sy, ey + one, sx, (ex - (p - one), ey - (p - one))),
        ),
    )

    result["TwoSum-SE-S4-X"] = z3.Implies(
        z3.And(same_sign, ex > ey, ex < ey + (p - two)),
        z3.Or(
            se_case_zero(sx, (ex, ex + one)),
            se_case(sx, ex, None, (ey - (p - one), ex - p)),
            se_case(sx, ex + one, None, (ey - (p - one), ex - (p - one))),
        ),
    )
    result["TwoSum-SE-S4-Y"] = z3.Implies(
        z3.And(same_sign, ey > ex, ey < ex + (p - two)),
        z3.Or(
            se_case_zero(sy, (ey, ey + one)),
            se_case(sy, ey, None, (ex - (p - one), ey - p)),
            se_case(sy, ey + one, None, (ex - (p - one), ey - (p - one))),
        ),
    )

    result["TwoSum-SE-S5"] = z3.Implies(
        z3.And(xy_nonzero, same_sign, ex == ey),
        z3.Or(
            se_case_zero(sx, ex + one),
            se_case(sx, ex + one, None, ex - (p - one)),
        ),
    )

    ###################################################### LEMMA FAMILY SE-D (5)

    # Lemmas in Family SE-D apply to addends with different signs.

    result["TwoSum-SE-D1-X"] = z3.Implies(
        z3.And(xy_nonzero, diff_sign, ex == ey + (p + one)),
        z3.Or(
            se_case(sx, ex - one, (sy,), (ey - (p - one), ex - (p + two))),
            z3.And(s_equals_x, e_equals_y),
        ),
    )
    result["TwoSum-SE-D1-Y"] = z3.Implies(
        z3.And(xy_nonzero, diff_sign, ey == ex + (p + one)),
        z3.Or(
            se_case(sy, ey - one, (sx,), (ex - (p - one), ey - (p + two))),
            z3.And(s_equals_y, e_equals_x),
        ),
    )

    result["TwoSum-SE-D2-X"] = z3.Implies(
        z3.And(diff_sign, ex == ey + p),
        z3.Or(
            se_case_zero(sx, ex - one),
            se_case(sx, ex - one, sy, (ey - (p - one), ex - (p + two))),
            se_case(sx, ex - one, (sy,), (ey - (p - one), ex - (p + one))),
            se_case(sx, ex, (sy,), (ey - (p - one), ex - p)),
            z3.And(s_equals_x, e_equals_y),
        ),
    )
    result["TwoSum-SE-D2-Y"] = z3.Implies(
        z3.And(diff_sign, ey == ex + p),
        z3.Or(
            se_case_zero(sy, ey - one),
            se_case(sy, ey - one, sx, (ex - (p - one), ey - (p + two))),
            se_case(sy, ey - one, (sx,), (ex - (p - one), ey - (p + one))),
            se_case(sy, ey, (sx,), (ex - (p - one), ey - p)),
            z3.And(s_equals_y, e_equals_x),
        ),
    )

    result["TwoSum-SE-D3-X"] = z3.Implies(
        z3.And(diff_sign, ex > ey + one, ex < ey + p),
        z3.Or(
            se_case_zero(sx, (ex - one, ex)),
            se_case(sx, ex - one, None, (ey - (p - one), ex - (p + one))),
            se_case(sx, ex, None, (ey - (p - one), ex - p)),
        ),
    )
    result["TwoSum-SE-D3-Y"] = z3.Implies(
        z3.And(diff_sign, ey > ex + one, ey < ex + p),
        z3.Or(
            se_case_zero(sy, (ey - one, ey)),
            se_case(sy, ey - one, None, (ex - (p - one), ey - (p + one))),
            se_case(sy, ey, None, (ex - (p - one), ey - p)),
        ),
    )

    result["TwoSum-SE-D4-X"] = z3.Implies(
        z3.And(diff_sign, ex == ey + one),
        z3.Or(
            se_case_zero(sx, (ex - p, ex)),
            se_case(sx, ex, None, ex - p),
        ),
    )
    result["TwoSum-SE-D4-Y"] = z3.Implies(
        z3.And(diff_sign, ey == ex + one),
        z3.Or(
            se_case_zero(sy, (ey - p, ey)),
            se_case(sy, ey, None, ey - p),
        ),
    )

    result["TwoSum-SE-D5"] = z3.Implies(
        z3.And(diff_sign, ex == ey),
        z3.Or(
            z3.And(s_pos_zero, e_pos_zero),
            se_case_zero(None, (ex - (p - one), ex - one)),
        ),
    )

    return result
